library/x509.c

```diff
@@ -780,10 +781,16 @@ int mbedtls_x509_dn_gets( char *buf, size_t size, const mbedtls_x509_name *dn )
 
             c = name->val.p[i];
             if( c < 32 || c == 127 || ( c > 128 && c < 160 ) )
-                 s[i] = '?';
-            else s[i] = c;
+                 s[i + j] = '?';
+            else if( c == ',' )
+            {
+                s[i + j] = '\\';
+                j++;
+                s[i + j] = c;
```

Comment:
I think there is a change this might overflow the buffer s. Note that in the previous version of the code at most 1 character was written per loop iteration, so the following check at the begining of each loop iteration was sufficient:
```
             if( i >= sizeof( s ) - 1 )
                 break;
```
I think that with the new version of the code, if that check passes, but there is only 1 byte of space left, then you will write 2 byte if the next character is a ,

Reply:
Good catch!
It should be:
~~~
if( i + j >= sizeof( s ) - 1 )
    break;
~~~

Reply:
Thanks for looking into it, but IMHO the new condition is not sufficient. For example, consider the case where at the beginning of an iteration in line 779 the condition `i + j >= sizeof( s ) - 1` is true but the condition `i + j + 1 >= sizeof( s ) - 1` is not. The check in line 779 will succeed. Then, if the character at `name->val.p[i]` is a `,` the program will eventually reach line 787 and the following two memory locations will be written:
```
s[i + j] = '\\';
s[i + j + 1] = c; 
```
Finally, in the following iteration the condition in line 779 will not be true and the location s[i + j] will be written with the value '\0', but at this point both i and j have been incremented by 1 so in fact the program is writing location s[i + j + 2] with regards to the values for i and j from the previous iteration causing a memory overwrite of 1 byte.

Reply:
Good point! I'll add additional checks for overflow